sky/metrics/utils.py

```diff
@@ -188,53 +188,61 @@ def start_svc_port_forward(context: str, namespace: str, service: str,
-
-        # read output line by line to find the local port
-        if port_forward_process.stdout:
-            # Wait up to 1s for data to be available without blocking
-            r, _, _ = select.select([port_forward_process.stdout], [], [],
-                                    _SELECT_TIMEOUT)
-            if r:
-                # Read available bytes from the FD without blocking
-                fd = port_forward_process.stdout.fileno()
-                raw = os.read(fd, _SELECT_BUFFER_SIZE)
-                chunk = raw.decode(errors='ignore')
-                buffer += chunk
-                match = re.search(r'Forwarding from 127\.0\.0\.1:(\d+)', buffer)
-                if match:
-                    local_port = int(match.group(1))
-                    break
-
-        # sleep for 100ms to avoid busy-waiting
-        time.sleep(0.1)
 
+    try:
+        # start the port forward process
+        port_forward_process = subprocess.Popen(cmd,
+                                                stdout=subprocess.PIPE,
+                                                stderr=subprocess.STDOUT,
+                                                text=True,
+                                                env=env)
+
+        start_time = time.time()
+
+        buffer = ''
+        # wait for the port forward to start and extract the local port
+        while time.time() - start_time < start_port_forward_timeout:
+            if port_forward_process.poll() is not None:
+                # port forward process has terminated
+                if port_forward_process.returncode != 0:
+                    port_forward_exit = True
+                break
+
+            # read output line by line to find the local port
+            if port_forward_process.stdout:
+                # Wait up to 1s for data to be available without blocking
+                r, _, _ = select.select([port_forward_process.stdout], [], [],
+                                        _SELECT_TIMEOUT)
+                if r:
+                    # Read available bytes from the FD without blocking
+                    fd = port_forward_process.stdout.fileno()
+                    raw = os.read(fd, _SELECT_BUFFER_SIZE)
+                    chunk = raw.decode(errors='ignore')
+                    buffer += chunk
+                    match = re.search(r'Forwarding from 127\.0\.0\.1:(\d+)',
+                                      buffer)
+                    if match:
+                        local_port = int(match.group(1))
+                        break
+
+            # sleep for 100ms to avoid busy-waiting
+            time.sleep(0.1)
+    except BaseException:  # pylint: disable=broad-exception-caught
+        if port_forward_process:
+            stop_svc_port_forward(port_forward_process,
+                                  timeout=terminate_port_forward_timeout)
+        raise
+    if port_forward_exit:
+        raise RuntimeError(f'Port forward failed for service {service} in '
```

Comment:
Should we call `stop_svc_port_forward` before raising the error?

Reply:
BTW, the errors raised in this function seem not handled in the caller function, should we catch the errors there and print logs so we know what happens later?